articles/151_roslaunch_xml.md

```diff
@@ -0,0 +1,255 @@
+Arguments are limited to the scope of their definition and thus have to be explictly passed to included files if any.
+
+##### Examples
+
+```xml
+<arg name="publish_frequency" default="10"/>
+<arg name="output_path" description="Output path for some processing pipeline"/>
+```
+
+#### `<node>` Tag
+
+##### Description
+
+The `<node>` tag allows for executing a ROS node in the form of local OS process.
+
+##### Examples
+
+```xml
+<node package="ros_demos" executable="publisher">
+  <param name="publish_frequency" value="10"/>
+  <remap from="generic_topic_name" to="my_topic"/>
+</node>
+```
+
+#### `<param>` Tag
+
+##### Description
+
+The `<param>` tag allows for setting ROS parameters of a ROS node.
+These can be scalar values or sequences of scalar values, defined directly or
+either nested or brought from a YAML file to make a map.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <param name="some_numeric_param" value="100.2"/>
+  <param name="some_list_of_strings" value="Some phrase,'100.0','true'" sep=","/>
+  <param name="some_list_of_ints" value="5, 3, 2" sep=", "/>
+  <param name="some_list_of_floats" value="5.0, 3.0, 2.0" sep=", "/>
+  <param name="some_param_group">
+    <param name="some_integer_param" value="10"/>
+  </param>
+  <param from="path/to/param/file.yml"/>
+</node>
+```
+
+#### `<remap>` Tag
+
+##### Description
+
+The `<remap>` tag allows for ROS name remapping of a `<node>` instance.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <remap from="chatter" to="/my_chatter"/>
+  <remap from="*/stuff" to="private_\1/stuff"/>
+</node>
+```
+
+#### `<env>` Tag
+
+##### Description
+
+The `<env>` tag allows for modifying an OS process environment.
+It can be used nested in `node` or `executable` tags.
+It doesn't allow conditional execution.
+
+##### Examples
+
+```xml
+<node package="my_pkg" executable="my_node">
+  <env name="RMW_IMPLEMENTATION" value="rmw_fastrtps_cpp"/>
+</node>
+
+<executable cmd="ls" cwd="/var/log">
+  <env name="LD_LIBRARY" value="/lib/some.so"/>
+</executable>
+```
+
+#### `<set_env>` Tag
+
+##### Description
+
+The `<set_env>` tag allows for modifying an OS process environment.
+It can be used nested in `launch` or `group` tags.
+It allows conditional execution.
+
+##### Examples
+
+```xml
+<group>
+  <set_env name="RMW_IMPLEMENTATION" value="rmw_fastrtps_cpp"/>
+</group>
+```
+
+#### `<unset_env>` Tag
+
+##### Description
+
+The `<unset_env>` tag allows for deleting an OS process environment variable.
+It can be used nested in `launch` or `group` tags.
+It allows conditional execution.
+
+##### Examples
+
+```xml
+<group>
+  <unset_env name="MY_ENV_VAR"/>
+</group>
+```
+
+## Dynamic Configuration
+
+### Substitution Syntax
+
+All substitutions are enclosed by `$(...)`.
+
+### Built-in Substitutions
+
+`$(find-package <pkg-name>)`
+: Substituted by the path to package installation directory in the local filesystem.
```

Comment:
What does "package installation directory" mean here? If it I the install prefix path of the package then I would suggest to use that term.

Reply:
It used this method, so yes @dirk-thomas's terminology is correct:

https://github.com/ament/ament_index/blob/0dba2f42c7f9b18a676c96d5643f26d5213513dd/ament_index_python/ament_index_python/packages.py#L38

It would be nice to also have a `$(find-package-share <pkg-name>)` as well as it would prevent the users from writing the package name twice, e.g.:

```
$(find-package really_long_package_name)/share/really_long_package_name/...
```

versus

```
$(find-package-share really_long_package_name)/...
```

Reply:
What are your thoughts on checking for file existance as part of a substitution? I've come across numerous instances where this would have been helpful over the years (e.g. large multi-robot project configuration where some robots override configuration and others don't).

Reply:
What are your thoughts on checking for file existance as part of a substitution? I've come across numerous instances where this would have been helpful over the years (e.g. large multi-robot project configuration where some robots override configuration and others don't).

Reply:
We could have a `$(find-file-in-package[-share] <pkg> <relative_path> [<default>])`, though we don't have one now. Or we could have a few small, sharp tools instead, like `$(file-exists $(find-package-share my_pkg)/some_launch.xml)`.